File: go/ppc/template.go
package capitan

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

type Variable = string
type Pattern = string
type Offset = int
type Kind uint8

const (
	Value Kind = iota
	Length
	Lines
	OffsetStart
	OffsetEnd
	LineStart
	LineEnd
	ColumnStart
	ColumnEnd
	FilePath
	FileName
	FileDirectory
	Lowercase
	Uppercase
	Capitalize
	Uncapitalize
	UpperCamelCase
	LowerCamelCase
	UpperSnakeCase
	LowerSnakeCase
	External
)

type TemplateElement struct {
	Constant string
	Hole     *Hole
}

type Hole struct {
	Pattern  Pattern
	Variable Variable
	Offset   Offset
	Kind     Kind
}

func attributeToKind(s string) Kind {
	switch s {
	case "value":
		return Value
	case "length":
		return Length
	case "lines":
		return Lines
	case "offset":
		return OffsetStart
	case "offset.start":
		return OffsetStart
	case "offset.end":
		return OffsetEnd
	case "line":
		return LineStart
	case "line.start":
		return LineStart
	case "line.end":
		return LineEnd
	case "column":
		return ColumnStart
	case "column.start":
		return ColumnStart
	case "column.end":
		return ColumnEnd
	case "file":
		return FilePath
	case "file.path":
		return FilePath
	case "file.name":
		return FileName
	case "file.directory":
		return FileDirectory
	case "lowercase":
		return Lowercase
	case "UPPERCASE":
		return Uppercase
	case "Capitalize":
		return Capitalize
	case "uncapitalize":
		return Uncapitalize
	case "UpperCamelCase":
		return UpperCamelCase
	case "lowerCamelCase":
		return LowerCamelCase
	case "UPPER_SNAKE_CASE":
		return UpperSnakeCase
	case "lower_snake_case":
		return LowerSnakeCase
	case "lsif.hover":
		return External
	default:
		panic("invalid attribute " + s)
	}
}

func camelToSnake(s string) string {
	var words []string
	var word []rune
	for _, r := range s {
		if isUpper(r) {
			if len(word) > 0 {
				words = append(words, strings.ToLower(string(word)))
				word = nil
			}
			word = append(word, unicode.ToLower(r))
		} else {
			word = append(word, r)
		}
	}
	if len(word) > 0 {
		words = append(words, strings.ToLower(string(word)))
	}
	return strings.Join(words, "_")
}

func substituteKind(filepath *string, hole *Hole, env map[Variable]string) (string, error) {
	var result string
	var err error

	switch hole.Kind {
	case Value:
		result, err = lookup(env, hole.Variable)
	case Length:
		result, err = length(env, hole.Variable)
	case Lines:
		result, err = lines(env, hole.Variable)
	case OffsetStart:
		result, err = offsetStart(env, hole.Variable)
	case OffsetEnd:
		result, err = offsetEnd(env, hole.Variable)
	case LineStart:
		result, err = lineStart(filepath, env, hole.Variable)
	case LineEnd:
		result, err = lineEnd(filepath, env, hole.Variable)
	case ColumnStart:
		result, err = columnStart(filepath, env, hole.Variable)
	case ColumnEnd:
		result, err = columnEnd(filepath, env, hole.Variable)
	case FilePath:
		if filepath != nil {
			result = *filepath
		}
	case FileName:
		if filepath != nil {
			result = filepath.Base(*filepath)
		}
	case FileDirectory:
		if filepath != nil {
			result = filepath.Dir(*filepath)
		}
	case Lowercase:
		result, err = lowercase(env, hole.Variable)
	case Uppercase:
		result, err = uppercase(env, hole.Variable)
	case Capitalize:
		result, err = capitalize(env, hole.Variable)
	case Uncapitalize:
		result, err = uncapitalize(env, hole.Variable)
	case UpperCamelCase:
		result, err = upperCamelCase(env, hole.Variable)
	case LowerCamelCase:
		result, err = lowerCamelCase(env, hole.Variable)
	case UpperSnakeCase:
		result, err = upperSnakeCase(env, hole.Variable)
	case LowerSnakeCase:
		result, err = lowerSnakeCase(env, hole.Variable)
	case External:
		result, err = external(filepath, env, hole.Variable)
	default:
		err = fmt.Errorf("unsupported kind %v", hole.Kind)
	}

	return result, err
}

func lookup(env map[Variable]string, variable Variable) (string, error) {
	value, ok := env[variable]
	if !ok {
		return "", fmt.Errorf("variable %q not found in environment", variable)
	}
	return value, nil
}

func length(env map[Variable]string, variable Variable) (string, error) {
	value, err := lookup(env, variable)
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("%d", len(value)), nil
}

func lines(env map[Variable]string, variable Variable) (string, error) {
	value, err := lookup(env, variable)
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("%d", strings.Count(value, "\n")+1), nil
}

func offsetStart(env map[Variable]string, variable Variable) (string, error) {
	return "0", nil // Placeholder implementation
}

func offsetEnd(env map[Variable]string, variable Variable) (string, error) {
	value, err := lookup(env, variable)
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("%d", len(value)), nil
}

func lineStart(filepath *string, env map[Variable]string, variable Variable) (string, error) {
	return "1", nil // Placeholder implementation
}

func lineEnd(filepath *string, env map[Variable]string, variable Variable) (string, error) {
	return "1", nil // Placeholder implementation
}

func columnStart(filepath *string, env map[Variable]string, variable Variable) (string, error) {
	return "1", nil // Placeholder implementation
}

func columnEnd(filepath *string, env map[Variable]string, variable Variable) (string, error) {
	return "1", nil // Placeholder implementation
}

func lowercase(env map[Variable]string, variable Variable) (string, error) {
	value, err := lookup(env, variable)
	if err != nil {
		return "", err
	}
	return strings.ToLower(value), nil
}

func uppercase(env map[Variable]string, variable Variable) (string, error) {
	value, err := lookup(env, variable)
	if err != nil {
		return "", err
	}
	return strings.ToUpper(value), nil
}

func capitalize(env map[Variable]string, variable Variable) (string, error) {
	value, err := lookup(env, variable)
	if err != nil {
		return "", err
	}
	return strings.Title(value), nil
}

func uncapitalize(env map[Variable]string, variable Variable) (string, error) {
	value, err := lookup(env, variable)
	if err != nil {
		return "", err
	}
	return strings
}

File: haskell/lib/Ppc/Template.hs
module Template (
    parseTemplate,
    variables,
    toString,
    substituteKind,
    attributeToKind
) where

import Control.Monad (guard)
import Data.Char (isAlphaNum, isUpper, isLower, toLower, toUpper)
import Data.List (intercalate)
import Data.Maybe (mapMaybe)
import System.Environment (lookupEnv)
import Text.Regex.Posix ((=~))

type Variable = String
type Pattern = String
type Offset = Int
type Kind = AttributeKind

data AttributeKind
    = Value
    | Length
    | Lines
    | OffsetStart
    | OffsetEnd
    | LineStart
    | LineEnd
    | ColumnStart
    | ColumnEnd
    | FilePath
    | FileName
    | FileDirectory
    | Lowercase
    | Uppercase
    | Capitalize
    | Uncapitalize
    | UpperCamelCase
    | LowerCamelCase
    | UpperSnakeCase
    | LowerSnakeCase
    | External String
    deriving (Show, Eq)

data TemplateElement
    = Constant String
    | Hole { pattern :: Pattern, variable :: Variable, offset :: Offset, kind :: Kind }
    deriving (Show)

debug :: Bool
debug = case lookupEnv "DEBUG_COMBY" of
    Nothing -> False
    Just _  -> True

attributeToKind :: String -> AttributeKind
attributeToKind = \case
    "value"          -> Value
    "length"         -> Length
    "lines"          -> Lines
    "offset"         -> OffsetStart
    "offset.start"   -> OffsetStart
    "offset.end"     -> OffsetEnd
    "line"           -> LineStart
    "line.start"     -> LineStart
    "line.end"       -> LineEnd
    "column"         -> ColumnStart
    "column.start"   -> ColumnStart
    "column.end"     -> ColumnEnd
    "file"           -> FilePath
    "file.path"      -> FilePath
    "file.name"      -> FileName
    "file.directory" -> FileDirectory
    "lowercase"      -> Lowercase
    "UPPERCASE"      -> Uppercase
    "Capitalize"     -> Capitalize
    "uncapitalize"   -> Uncapitalize
    "UpperCamelCase" -> UpperCamelCase
    "lowerCamelCase" -> LowerCamelCase
    "UPPER_SNAKE_CASE" -> UpperSnakeCase
    "lower_snake_case" -> LowerSnakeCase
    "lsif.hover"     -> External "lsif.hover"
    s                -> error $ "invalid attribute " ++ show s

parseTemplate :: String -> [TemplateElement]
parseTemplate = undefined

variables :: [TemplateElement] -> [TemplateElement]
variables = mapMaybe $ \case
    Hole h -> Just h
    _      -> Nothing

toString :: [TemplateElement] -> String
toString = concatMap $ \case
    Constant c -> c
    Hole { pattern } -> pattern

camelToSnake :: String -> String
camelToSnake = intercalate "_" . go 0 . map (\case
    c | isUpper c -> ['_', toLower c]
    c             -> [c])
  where
    go _ []     = []
    go n (x:xs) | isUpper x = x : go (n + 1) xs
                | otherwise = toLower x : go (n + 1) xs

substituteKind :: Maybe FilePath -> TemplateElement -> Environment -> Maybe String
substituteKind filepath (Hole { variable, kind }) env = case kind of
    Value         -> lookup env variable
    Length        -> fmap show . fmap length $ lookup env variable
    Lines         -> fmap (show . (+ 1) . length . filter (== '\n')) $ lookup env variable
    OffsetStart   -> fmap show . matchStart env variable
    OffsetEnd     -> fmap show . matchEnd env variable
    LineStart     -> lineStart filepath env variable
    LineEnd       -> lineEnd filepath env variable
    ColumnStart   -> columnStart filepath env variable
    ColumnEnd     -> columnEnd filepath env variable
    FilePath      -> filepath
    FileName      -> fmap fileName filepath
    FileDirectory -> fmap directory filepath
    Lowercase     -> fmap map toLower $ lookup env variable
    Uppercase     -> fmap map toUpper $ lookup env variable
    Capitalize    -> fmap (let (h:t) = id in toUpper h : map toLower t) $ lookup env variable
    Uncapitalize  -> fmap (let (h:t) = id in toLower h : t) $ lookup env variable
    UpperCamelCase -> fmap (concatMap capitalize . words . camelToSnake) $ lookup env variable
    LowerCamelCase -> fmap (let (h:t) = concat . map capitalize . words . camelToSnake in toLower h : t) $ lookup env variable
    UpperSnakeCase -> fmap (map toUpper . camelToSnake) $ lookup env variable
    LowerSnakeCase -> fmap camelToSnake $ lookup env variable
    External "lsif.hover" -> do
        filepath' <- filepath
        value <- lookup env variable
        offset <- matchStart env variable
        let source = readFile filepath'
            (line, column) = offsetToLineColumn source offset
        pure $ "lsif.hover for " ++ show (line, column) ++ ": " ++ value
    External _ -> error "Unsupported external attribute"
  where
    lookup env var = flip lookup env <$> stripVar var
    matchStart env var = fmap fst . matchRange env <$> stripVar var
    matchEnd env var = fmap snd . matchRange env <$> stripVar var
    lineStart filepath env var = do
        filepath' <- filepath
        offset <- matchStart env var
        let source = readFile filepath'
            (line, _) = offsetToLineColumn source offset
        pure $ show line
    lineEnd filepath env var = do
        filepath' <- filepath
        offset <- matchEnd env var
        let source = readFile filepath'
            (line, _) = offsetToLineColumn source offset
        pure $ show line
    columnStart filepath env var = do
        filepath' <- filepath
        offset <- matchStart env var
        let source = readFile filepath'
            (_, column) = offsetToLineColumn source offset
        pure $ show column
    columnEnd filepath env var = do
        filepath' <- filepath
        offset <- matchEnd env var
        let source = readFile filepath'
            (_, column) = offsetToLineColumn source offset
        pure $ show column
    fileName = takeWhileEnd (/= '/') . takeWhileEnd (/= '\\')
    directory = dropWhileEnd (/= '/') . dropWhileEnd (/= '\\')
    offsetToLineColumn source offset = (1, 1) -- Placeholder implementation
    stripVar var = case var =~ "^\\[(.*?)\\]$" of
        Just [_, v] -> Just v
        _           -> Nothing

-- Placeholder implementation
readFile :: FilePath -> String
readFile _ = ""

matchRange :: Environment -> Variable -> (Int, Int)
matchRange _ _ = (0, 0)

type Environment = [(Variable, String)]